src/Session.cpp

```diff
@@ -1102,37 +1099,55 @@ ISampleReader* CSession::GetNextSample()
     if (!streamReader)
       continue;
 
-    if (stream->m_isEnabled && !streamReader->EOS() &&
-        AP4_SUCCEEDED(streamReader->Start(isStarted)))
+    if (stream->m_isEnabled)
     {
-      if (!res || streamReader->DTSorPTS() < res->GetReader()->DTSorPTS())
+      // Advice is that VP does not want to wait longer than 10ms for a return from
+      // DemuxRead() - here we ask to not wait at all and if ReadSample has not yet
+      // finished we return the dummy reader instead
+      if (streamReader->IsReadSampleAsyncWorking())
       {
-        if (stream->m_adStream.waitingForSegment(true))
-          waiting = stream.get();
-        else
-          res = stream.get();
+        waiting = stream.get();
+        break;
+      }
+      else if (!streamReader->EOS() &&
+        AP4_SUCCEEDED(streamReader->Start(isStarted)))
+      {
+        if (!res || streamReader->DTSorPTS() < res->GetReader()->DTSorPTS())
+        {
+          if (stream->m_adStream.waitingForSegment(true))
+          {
+            waiting = stream.get();
+          }
+          else
+          {
+            res = stream.get();
+          }
+        }
       }
     }
 
     if (isStarted && streamReader->GetInformation(stream->m_info))
       m_changed = true;
   }
 
-  if (res)
+  if (waiting)
   {
-    CheckFragmentDuration(*res);
-    if (res->GetReader()->GetInformation(res->m_info))
-      m_changed = true;
-    if (res->GetReader()->PTS() != STREAM_NOPTS_VALUE)
-      m_elapsedTime = PTSToElapsed(res->GetReader()->PTS()) + GetChapterStartTime();
-    return res->GetReader();
+    return true;
   }
-  else if (waiting)
+  else if (res)
   {
-    std::this_thread::sleep_for(std::chrono::milliseconds(100));
-    return m_dummySampleReader.get();
+    CheckFragmentDuration(*res);
+    ISampleReader* sr{res->GetReader()};
```

Comment:
Better to use `ISampleReader* sr = stream->GetReader();` No need to a different assignment operator that is not often used.